src/s3/util/s3Tab.ts

```diff
@@ -0,0 +1,103 @@
+/*!
+ * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as vscode from 'vscode'
+//import { Bucket, DownloadFileRequest, File, S3Client } from '../../shared/clients/s3Client'
+
+//need to subscribe to ondidsave and upload to s3, might need the S3FileNode
+
+export class S3Tab {
+    //private file: File
+    private s3Uri: vscode.Uri
+    private window: typeof vscode.window
+    private editor: vscode.TextEditor | undefined
+
+    //private editing: boolean
+
+    public constructor(private fileUri: vscode.Uri, window = vscode.window) {
+        this.s3Uri = vscode.Uri.parse('s3:' + this.fileUri.fsPath)
+        this.window = window
+    }
+
+    public async openFileOnReadOnly(workspace = vscode.workspace): Promise<vscode.TextEditor | undefined> {
+        //find if there is any active editor for this uri
+        const openEditor = await this.getActiveEditor()
+
+        if (!openEditor) {
+            //there isn't any tab open for this uri, simply open it in read-only with the s3Uri
+            const doc = await workspace.openTextDocument(this.s3Uri)
+            this.editor = await this.window.showTextDocument(doc, { preview: false })
+
+            return this.editor
+        } else if (openEditor.document.uri.scheme === 's3') {
+            //there is one already in read-only, just shift focus to it by reopening it with the ViewColumn option
+            await this.window.showTextDocument(openEditor.document, {
+                preview: false,
+                viewColumn: openEditor.viewColumn,
+            })
+            this.editor = openEditor
+
+            return this.editor
+        } else if (openEditor.document.uri.scheme === 'file') {
+            //there is a tab open in edit-mode, it needs to be focused, then closed
+            await this.window.showTextDocument(openEditor.document, {
+                preview: false,
+                viewColumn: openEditor.viewColumn,
+            })
+            await vscode.commands.executeCommand('workbench.action.closeActiveEditor')
+            //good to open in read-only
+            const doc = await workspace.openTextDocument(this.s3Uri)
+            this.editor = await this.window.showTextDocument(doc, { preview: false })
+
+            return this.editor
```

Comment:
Not sure if there is really a need to revert back into the read only mode if I already have an edit mode tab open. The only case in which I see this being useful is if I wanted to re-preview the latest version of the file to see if it's been modified since I opened it. There is probably a better way to accomplish this though. (diff? check on save attempts?)

Reply:
I think it was Jo's suggestion to reopen in read-only when saving, for safety

Reply:
OK - I thought this particular logic was switching back to preview mode simply by re-selecting the file. If we do so only on save I'm OK with that.

Reply:
Nvm I thought this comment was regarding the reopen on read-only after saving, it is also possible to switch to read-only by reclicking the `s3FileNode`

Reply:
IMO: keep the former, ditch the latter. 😄

Reply:
Addressed with: https://github.com/aws/aws-toolkit-vscode/pull/1878/files#diff-533c6b5e69cbb4ec72bc424ef27945bfb5355535c4e354450753051879dcdc6dR47 and